### PR TITLE
Honor user setting overrides for current culture when using ICU on Windows

### LINKED ISSUE #37121: Enable Globalization to read the user setting overrides for current cultures when using ICU
### Description
The **[DateTimeFormatInfo](https://docs.microsoft.com/en-us/dotnet/api/system.globalization.datetimeformatinfo?view=netcore-3.1)** properties in Net Core 5 Preview 4 return different culture-specific information about the format of date and time values compared to .Net Framework in the same environment

**Our unit tests detected this issue immediately.** (we maintain unit tests for all the released and preview environments like .net core 5 in our codebase). Thus it is an **important breaking change for us**.

### Configuration
OS Version: Microsoft Windows [Version 10.0.18362.535]
.Net Framework Version:   Microsoft (R) .NET CLR Version Tool  Version 4.8.3928.0
.Net Core Version: SDK Version: 5.0.100-preview.4.20258.7

### Regression?
Yes, this works correctly until .Net Core 5 Preview 4

### Other information
Here are some tests which fail under  .Net Core 5 Preview 4
```cs
[TestCase("en-US", "M/d/yyyy", "dddd, MMMM d, yyyy")]
[TestCase("ru-RU", "dd.MM.yyyy", "d MMMM yyyy 'г.'")]
public void DatePatterns(string culture, string pShort, string pLong) {
	CultureInfo c = new CultureInfo(culture);
	Assert.AreEqual(pShort, c.DateTimeFormat.ShortDatePattern);
	Assert.AreEqual(pLong, c.DateTimeFormat.LongDatePattern);
}
[TestCase("en-US", "h:mm tt", "h:mm:ss tt")]
[TestCase("ru-RU", "H:mm", "H:mm:ss")]
public void TimePatterns(string culture, string pShort, string pLong) {
	CultureInfo c = new CultureInfo(culture);
	Assert.AreEqual(pShort, c.DateTimeFormat.ShortTimePattern);
	Assert.AreEqual(pLong, c.DateTimeFormat.LongTimePattern);
}
[TestCase("en-US", "January", "Jan")]
[TestCase("ru-RU", "Январь", "янв")]
public void MonthNames(string culture, string name, string abrName) {
	CultureInfo c = new CultureInfo(culture);
	Assert.AreEqual(name, c.DateTimeFormat.MonthNames[0]);
	Assert.AreEqual(abrName, c.DateTimeFormat.AbbreviatedMonthNames[0]);
}
[TestCase("en-US", "January", "Jan")]
[TestCase("ru-RU", "января", "янв")]
public void MonthGenitiveNames(string culture, string name, string abrName) {
	CultureInfo c = new CultureInfo(culture);
	Assert.AreEqual(name, c.DateTimeFormat.MonthGenitiveNames[0]);
	Assert.AreEqual(abrName, c.DateTimeFormat.AbbreviatedMonthGenitiveNames[0]);
}
```

### PR DESCRIPTION
**Pull request by safern**
Fixes: https://github.com/dotnet/runtime/issues/37121

Adding support to read user overrides from the OS when using ICU on Windows when the locale information we're getting is the user's default locale.

@tarekgh the only way I could think we could test this would be by modifying the registry key and then clearing the cache data, do you think we should do that?

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Unix.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Unix.cs:1]: License header

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.Windows.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.Windows.cs:47]: Delete copy of this method in CalendarData.Nls.cs ?

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.cs:102]: >bool bUseUserOverrides, bool isUserDefaultLocale [](start = 72, length = 48)

can we combine these 2 flags. I mean sending bUseUserOverrides = true only when it is user default locale?

Comment (safern) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.cs:102]: Yes we can do that, that would affect NLS though but seems reasonable to only ask for user overrides when we're in the default culture.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:1412]: >(user can override)  [](start = 12, length = 20)

please move this comment to its own line.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:1442]: >GetTimeFormatsCore [](start = 26, length = 18)

should we have 2 copies of such methods one for Windows and one for Linux? just to avoid the ShouldUseUserOverrideNlsData check on Linux and to ensure running the rest of the code on Windows only. 

otherwise we can keep this way but adding assert after this check to ensure we are running on Windows.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:1397]: >GetTimeFormatsCore [](start = 42, length = 18)

I think the user can override the default time format and not all time formats. in other word, we need to read the whole list from ICU and the default one from NLS. this will apply for all properties returning a list.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs
Comment (safern) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:1442]: Yeah we can do that.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs
Comment (safern) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:1412]: This is how we have it in all properties where the user can override. Should we do the same for all those?

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs
Comment (safern) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:1397]: Ok, makes sense.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs
Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:1412]: It is confusing to have written this way. but it is not that important. I'll leave it to you :-)

---
In reply to: [445747823](https://github.com/dotnet/runtime/pull/38372#discussion_r445747823) [](ancestors = 445747823)

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:1710]: >ShouldUseUserOverrideNlsData [](start = 32, length = 28)

this will be same too. getting the list from ICU and getting the default from NLS. later in this method we set the default. so I guess in this place we always read the list without caring about user overrides and then later when setting the default we do the user overrides check.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:1887]: >NlsGetLocaleInfo [](start = 71, length = 16)

for codepages, we can always read them from NLS if we are running on Windows OS. users cannot override these and ICU doesn't provide them. We carry our own data to use when running on Linux/OSX. this comment apply to the 4 codepages.

Comment (safern) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:1887]: Ok, will do that.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Windows.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Windows.cs:117]: we need to apply same logic for these arrays too. read from ICU the list and read the default from NLS.

Comment (safern) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Windows.cs:117]: Ok, makes sense.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Windows.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Windows.cs:83]: >CultureInfo.GetUserDefaultLocaleName(); [](start = 52, length = 39)

should we use the cached value s_userDefaultLocaleName instead? I think we shouldn't call the OS every time we create a culture.

Comment (safern) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Windows.cs:83]: Will it call the OS everytime? `CultureInfo.GetUserDefaultLocaleName` is using `Interlocked.CompareExchange` and only calling the OS when `s_userDefaultLocaleName == null`.

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Windows.cs:83]: also we need to do that only if useroverride is true. that is why I was preferring to have just one variable encapsulate the logic.

---
In reply to: [445776569](https://github.com/dotnet/runtime/pull/38372#discussion_r445776569) [](ancestors = 445776569)

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.cs:754]: >null [](start = 38, length = 4)

we should just initialize it here. Also we should initialize it at its definition.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Windows.cs
Comment (safern) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Windows.cs:83]: > also we need to do that only if useroverride is true

You mean, only do the caching if `userOverride` is true, or doing the comparison at all?

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.Windows.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.Windows.cs:47]: >s_userDefaultLocaleName [](start = 40, length = 23)

should just initialize it here.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Windows.cs
Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Windows.cs:83]: I mean encapsulate the userOverride and _isDefaultUserLocale to only one variable. and this check will be like 

```C#
userOverride = userOverride || _sWindowsName == CultureInfo.GetUserDefaultLocaleName();
```

or something like that.

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Windows.cs:83]: > Will it call the OS everytime? CultureInfo.GetUserDefaultLocaleName is using Interlocked.CompareExchange and only calling the OS when s_userDefaultLocaleName == null.

That will be too much logic calling compare exchange every time you need to access the variable. I don't think we need to do that. caching the value in the initialized static variable should be enough and simpler.

Comment (safern) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Windows.cs:83]: Makes sense... working on that change now

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs
Comment (iSazonov) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:1397]: Will an running application see new time format if user changes the time format on the fly?

Comment (safern) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:1397]: It will have the same behavior as it has now, currently we already cache the values so that wouldn’t be a different behavior.

I think applications need to call CultureInfo.ClearCache().

Comment (sba923) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:1397]: I would expect that PowerShell (e.g. Get-ChildItem's console output, currently affected by this issue) takes user changes into account on the fly.

Comment (iSazonov) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:1397]: @safern Do you say about ICU only behavior? Was there a cache for NLS too (in .Net 3)?

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:1397]: There is no change in the behavior than before here. The default user locale was cached before and can be reset by calling ClearCachedData. 
@sba923 @iSazonov I would suggest after @safern merge this change, you give it a try and let's know if you see any regression. We'll be happy to address it.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Windows.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Windows.cs:68]: >InsertValueToArray [](start = 21, length = 18)

does this need to check if the value we are adding is already exist in the array before we add it?

Comment (safern) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Windows.cs:68]: Good idea, maybe we should insert if it doesn't exist but if it exists move it to index[0].

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Windows.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Windows.cs:70]: nit: you could just do if `(value == null) return;` this will avoid including the whole block inside a parenthesis 

just a suggestion  you don't have to do it.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Windows.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Windows.cs:77]: >index = i; [](start = 24, length = 10)

nit: you can simplify this by just doing the swap here and then return from the method. this will get rid of the index variable and avoid more checking later and should make the code much simpler. 

just a suggestion  you don't have to do it.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Windows.cs
Comment (safern) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Windows.cs]: Thanks, fixed.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Windows.cs
Comment (safern) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Windows.cs]: Thanks, fixed.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Unix.cs
Comment (danmosemsft) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Unix.cs:84]: Aside (not this PR) it would be nice at some point to remove all the ugly Hungarian prefixes in CultureData

Comment (safern) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Unix.cs:84]: Yeah, I agree, that'd be a nice cleanup.

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Unix.cs:84]: Hungarian notations is actually to match Windows exposed API's fields. I am not pushing again the clean up but wanted to tell why we had it in the first place.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.Windows.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.Windows.cs:45]: >private [](start = 61, length = 7)

why we have it private set if it is internal property anyway?

**Review by tarekgh — APPROVED** [review]
Thanks @safern for getting this fixed

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.Windows.cs
Comment (safern) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.Windows.cs:45]: We should just be setting it from within `CultureInfo` -- I can remove the `private set`.

Comment (safern) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.Windows.cs:45]: I'll remove it, I just realized I'll have to update the PR, because in `CultureInfo.ClearChache()` I'm not using InvariantCulture.Name when in invariant mode.

Comment (safern) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.Windows.cs]: Done.

**Pull request merged**
